From fbe6d654db34edf8bf2c12b29b1a9298ca2d21b4 Mon Sep 17 00:00:00 2001
From: Che-Wei Chuang <cnoize@andestech.com>
Date: Mon, 13 Jan 2025 11:26:00 +0800
Subject: [PATCH 34/53] swdtb: uart: update current-speed property based on
 CONFIG_BAUDRATE (#136)

Update the UART nodes in the device tree to reflect the value of CONFIG_BAUDRATE.
If the current-speed property already exists, its value is updated to match CONFIG_BAUDRATE.
If the property does not exist, it is added with the correct value.

This current-speed property will be passed to OpenSBI to configure the baud rate
when it displays messages, ensuring accurate UART communication during system initialization.

Signed-off-by: Che-Wei Chuang <cnoize@andestech.com>

Reviewed-on: https://gitea.andestech.com/RD-SW/u-boot.official/pulls/136
Reviewed-by: randolph <randolph@andestech.com>
Reviewed-by: Leo Yu-Chi Liang <ycliang@andestech.com>
Reviewed-by: Tim Shih-Ting OuYang <tim609@andestech.com>
Co-authored-by: Che-Wei Chuang <cnoize@andestech.com>
Co-committed-by: Che-Wei Chuang <cnoize@andestech.com>
---
 .../common/swdtb_merger/swdtb_merger.c        |   2 +
 .../common/swdtb_merger/swdtb_merger_fixup.c  | 100 +++++++++++++++++-
 .../common/swdtb_merger/swdtb_merger_fixup.h  |   1 +
 3 files changed, 98 insertions(+), 5 deletions(-)

diff --git a/board/andestech/common/swdtb_merger/swdtb_merger.c b/board/andestech/common/swdtb_merger/swdtb_merger.c
index cedbc7384681..4992b7d15ed4 100644
--- a/board/andestech/common/swdtb_merger/swdtb_merger.c
+++ b/board/andestech/common/swdtb_merger/swdtb_merger.c
@@ -59,6 +59,8 @@ struct swdtb_table dtb_merge_table[] = {
 	 "add dmas",			mmc_dma_dmas_fixup},
 	{SWDTB_ENABLE,	1,
 	 "add dma cell",		dma_cells_fixup},
+	{SWDTB_ENABLE,	1,
+	 "fix uart baud rate",		uart_baudrate_fixup},
 };
 
 /*
diff --git a/board/andestech/common/swdtb_merger/swdtb_merger_fixup.c b/board/andestech/common/swdtb_merger/swdtb_merger_fixup.c
index da15fcc3c6f5..a846ff2c0011 100644
--- a/board/andestech/common/swdtb_merger/swdtb_merger_fixup.c
+++ b/board/andestech/common/swdtb_merger/swdtb_merger_fixup.c
@@ -151,8 +151,8 @@ int riscv_isa_fixup(void *fdt, int node, char *name, char *value)
 	int ret;
 	/*
 	 * Cannot detect the old string, the old string is a
-	 * subset of the new string. 
-	 * 0 is no new_dsp, 1 indicates that it contains new_dsp. 
+	 * subset of the new string.
+	 * 0 is no new_dsp, 1 indicates that it contains new_dsp.
 	 */
 	int new_dsp_idx = 0;
 	int old_xv5_idx = 0;
@@ -260,7 +260,7 @@ int riscv_isa_fixup(void *fdt, int node, char *name, char *value)
 			dtb_error("Error on fdt_setprop: %s\n",
 				  fdt_strerror(ret));
 			return ret;
- 		}
+		}
 
 		show_result(fdt, node, prop_name, RESULT_STRING);
 
@@ -789,7 +789,7 @@ int cpu_intc_fixup(void *fdt, int node, char *name, char *value)
 	char *new_str = "andestech,cpu-intc";
 	int type = COMPAT_TYPE_APPEND;
 
-	return compatible_change(fdt, node, name, value, 
+	return compatible_change(fdt, node, name, value,
 				 old_str, new_str, type);
 }
 
@@ -924,5 +924,95 @@ int dma_cells_fixup(void *fdt, int node, char *name, char *value)
 		}
 	}
 
-	return !(RETURN_DONE);	
+	return !(RETURN_DONE);
+}
+
+/**
+ * uart_baudrate_fixup() modifies or adds the `current-speed` property
+ * to all UART nodes in the DTS using the CONFIG_BAUDRATE value.
+ *
+ * DTS:
+ *   serial@f0300000 {
+ *     compatible = "andestech,uart16550\0ns16550a";
+ *     reg = <0x00 0xf0300000 0x00 0x1000>;
+ *     interrupts = <0x09 0x04>;
+ *     interrupt-parent = <0x03>;
+ *     clock-frequency = <0x12c0000>;
+ *     current-speed = <38400>;
+ *     reg-shift = <0x02>;
+ *     reg-offset = <0x20>;
+ *     reg-io-width = <0x04>;
+ *     no-loopback-test = <0x01>;
+ *   };
+ *
+ * If CONFIG_BAUDRATE==115200, the function ensures:
+ *   current-speed = <115200>;
+ *
+ * @return 0 if successful, or FDT_ERR... if not.
+ */
+int uart_baudrate_fixup(void *fdt, int node, char *name, char *value)
+{
+	const char *uart_compatible = "andestech,uart16550";
+	const char *current_speed_name = "current-speed";
+	int node_offset = -1;
+	int ret;
+	int config_baudrate = CONFIG_BAUDRATE;
+
+	/* Iterate through all nodes to find UART nodes */
+	while ((node_offset = fdt_node_offset_by_compatible(
+			fdt, node_offset, uart_compatible)) >= 0) {
+		const u32 *current_speed;
+		int len;
+
+		/* Check if `current-speed` already exists */
+		current_speed = fdt_getprop(fdt, node_offset,
+					    current_speed_name, &len);
+		if (current_speed) {
+			dtb_info("Node: %s, Current-speed: %d\n",
+				 fdt_get_name(fdt, node_offset, NULL),
+				 *current_speed);
+
+			/* Skip if speed matches config */
+			if (*current_speed == config_baudrate)
+				continue;
+
+			/* If it exists, update the value */
+			ret = fdt_setprop_u32(fdt, node_offset,
+					      current_speed_name,
+					      config_baudrate);
+			if (ret) {
+				dtb_error("Failed to update %s: %s\n",
+					  current_speed_name,
+					  fdt_strerror(ret));
+				return ret;
+			}
+		} else {
+			/* If it does not exist, add the property */
+			ret = fdt_appendprop_u32(fdt, node_offset,
+						 current_speed_name,
+						 config_baudrate);
+			if (ret) {
+				dtb_error("Failed to add %s: %s\n",
+					  current_speed_name,
+					  fdt_strerror(ret));
+				return ret;
+			}
+			dtb_info("Node: %s, Current-speed: %d\n",
+				 fdt_get_name(fdt, node_offset, NULL),
+				 config_baudrate);
+		}
+
+		/* Debug output to show the result */
+		show_result(fdt, node_offset, current_speed_name,
+			    RESULT_NUMERAL);
+	}
+
+	/* If no UART nodes were found, return an error */
+	if (node_offset < 0 && node_offset != -FDT_ERR_NOTFOUND) {
+		dtb_error("Error while searching for UART nodes: %s\n",
+			  fdt_strerror(node_offset));
+		return node_offset;
+	}
+
+	return RETURN_DONE;
 }
diff --git a/board/andestech/common/swdtb_merger/swdtb_merger_fixup.h b/board/andestech/common/swdtb_merger/swdtb_merger_fixup.h
index 7966a486bc89..38ecc9aad323 100644
--- a/board/andestech/common/swdtb_merger/swdtb_merger_fixup.h
+++ b/board/andestech/common/swdtb_merger/swdtb_merger_fixup.h
@@ -58,4 +58,5 @@ int andespmu_event_fixup(void *fdt, int node, char *name, char *value);
 int mmc_dma_rxtx_fixup(void *fdt, int node, char *name, char *value);
 int mmc_dma_dmas_fixup(void *fdt, int node, char *name, char *value);
 int dma_cells_fixup(void *fdt, int node, char *name, char *value);
+int uart_baudrate_fixup(void *fdt, int node, char *name, char *value);
 #endif
-- 
2.34.1

